File: GuardianX-main/backend/app/schemas/schedule.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

ScheduleCadence = Literal["DAILY", "WEEKLY", "MONTHLY"]
WeekDay = Literal["MON", "TUE", "WED", "THU", "FRI", "SAT", "SUN"]


class ScheduledScanCreate(BaseModel):
    asset_id: int
    cadence: ScheduleCadence
    time_of_day: str = Field(
        pattern=r"^([01]\d|2[0-3]):[0-5]\d$",
        description="UTC time of day, format HH:MM",
    )
    week_day: WeekDay | None = None
    month_day: int | None = Field(
        default=None,
        ge=1,
        le=31,
    )
    scanner: str = "nmap"
    enabled: bool = True


class ScheduledScanUpdate(BaseModel):
    cadence: ScheduleCadence | None = None
    time_of_day: str | None = Field(
        default=None,
        pattern=r"^([01]\d|2[0-3]):[0-5]\d$",
    )
    week_day: WeekDay | None = None
    month_day: int | None = Field(
        default=None,
        ge=1,
        le=31,
    )
    scanner: str | None = None
    enabled: bool | None = None


class ScheduledScanResponse(BaseModel):
    id: int
    asset_id: int
    asset_name: str | None = None
    scanner: str
    cadence: str
    time_of_day: str
    week_day: str | None
    month_day: int | None
    enabled: bool
    last_run_at: datetime | None
    next_run_at: datetime | None
    created_by: int
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(
        from_attributes=True,
    )

File: GuardianX-main/guardianx-frontend/src/types/schedule.ts
export type ScheduleCadence = "DAILY" | "WEEKLY" | "MONTHLY";

export type ScheduleWeekDay =
  | "MON"
  | "TUE"
  | "WED"
  | "THU"
  | "FRI"
  | "SAT"
  | "SUN";

export interface ScheduledScan {
  id: number;
  asset_id: number;
  asset_name: string | null;
  scanner: string;
  cadence: ScheduleCadence;
  time_of_day: string;
  week_day: ScheduleWeekDay | null;
  month_day: number | null;
  enabled: boolean;
  last_run_at: string | null;
  next_run_at: string | null;
  created_by: number;
  created_at: string;
  updated_at: string;
}

export interface CreateScheduleDto {
  asset_id: number;
  cadence: ScheduleCadence;
  time_of_day: string;
  week_day?: ScheduleWeekDay | null;
  month_day?: number | null;
  scanner?: string;
  enabled?: boolean;
}

export type UpdateScheduleDto = Partial<CreateScheduleDto>;
